### ISSUE TITLE
Mouse events have latency on macos due to a change in accesskit crate

### ISSUE THREAD

**Issue by artursapek** [bevyengine/bevy#9391]
## Bevy version

0.11 (anything after https://github.com/bevyengine/bevy/pull/8655)

## Relevant system information

macos `13.5 (22G74)` on 2020 M1 Macbook Air

## What you did

I'm running code that uses mouse events, such as the `mouse_input` example

## What went wrong

There is a noticeable delay in mouse clicks being captured. See https://imgur.com/a/py58Zk4. On the left is bevy 0.11, and on the right is bevy 0.10. Watch with the sound on to hear when I click my mouse.

## Additional information

I binary searched through the commit log until I was able to narrow it down to this specific change inside `accesskit`. https://github.com/AccessKit/accesskit/pull/231/files#r1287857891

If I revert that code change, the observed latency in bevy capturing mouse events disappears.

I don't know what to do here. I don't understand why that change was made to `accesskit`, or exactly why it's causing this latency. I am an end-user of bevy who got annoyed and jumped down this rabbit hole and am looking for an opinion from bevy core devs. It seems like accessibility features slowing down an example as simple as `mouse_input` indicate the accessibility code might need to be isolated better.

**Comment by mwcampbell** [comment]
Does this affect all users of Bevy-based games on macOS?

@artursapek I gather from the video that you're not using VoiceOver. Do you have any other macOS accessibility features enabled? AccessKit should have no effect on machines that aren't running any accessibility tools (a.k.a. assistive technologies).

**Comment by artursapek** [comment]
@mwcampbell I went through them all in system prefs and none appear to be enabled. is there a way for me to add a print into `accesskit` that will tell me the reason it's enabled?

**Comment by mwcampbell** [comment]
As far as I know, there's no way for AccessKit to know which accessibility client is calling it. Maybe someone who knows much more about macOS internals than I do could come up with some kind of hack to determine that. But I think it's much more important to find out how widespread the problem reported in this issue is.

**Comment by artursapek** [comment]
I agree. I don't know how widespread it is. I haven't tested it on a second Mac yet. hoping other people chime in here.

**Comment by basilkohler** [comment]
Wanted to report the same issue: Left click has a noticeable delay.
M1 Macbook pro 16 on latest MacOS 13.5.
Right click works as expected.

**Comment by mwcampbell** [comment]
I think we still need an answer to the question of whether this is affecting all macOS users or only a minority. If the latter, that doesn't mean it's unimportant, of course. But in that case, it may help to figure out what the common factor is. This question also affects the urgency of the issue.

**Comment by artursapek** [comment]
I'm really not sure, but I would expect more people to be complaining if this affected all macos users.

regardless though, it's slow code and should be fixed

**Comment by aevyrie** [comment]
Just noticed this as well, only seems to impact LMB. VoiceOver is disabled.

**Comment by aevyrie** [comment]
It looks like `Res<bevy::a11y::AccessibilityRequested>` is being set to `true` as soon as I click on the window, even though I'm not using any assistive technologies. However, inputs shouldn't be hanging regardless.

This is the behavior I am noticing:

```
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 7.871041ms
[crates/bevy_picking_input/src/mouse.rs:49] pointer_dist = 0.0
[crates/bevy_picking_input/src/mouse.rs:73] "down" = "down"
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 17.210834ms
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 7.775917ms
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 16.602542ms
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 8.354875ms
[crates/bevy_picking_input/src/mouse.rs:49] pointer_dist = 435.2715
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 17.234292ms
[crates/bevy_picking_input/src/mouse.rs:49] pointer_dist = 11.138973
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 7.884667ms
[crates/bevy_picking_input/src/mouse.rs:49] pointer_dist = 13.220875
[crates/bevy_picking_input/src/mouse.rs:82] frametime = 16.808792ms
```

The frame time is printed every frame. The pointer dist is how far the pointer travelled according to bevy's inputs. The pattern I see is that on the frame where a mouse down is received, first bevy sends a mousse delta of zero, then it stops sending mouse moves entirely for a while, then suddenly will send a huge jump (see the `pointer_dist = 435.2715` above).

**Comment by aevyrie** [comment]
I've confirmed this is **not** the issue, as initially suspected: https://github.com/bevyengine/bevy/blob/22e39c4abf6e2fdf99ba0820b3c35db73be71347/crates/bevy_winit/src/lib.rs#L508-L512

**Comment by aevyrie** [comment]
Another observation - running with vsync off masks the issue, suggesting the delay isn't related to some call taking some amount of time, rather, it seems to be delayed by some number of frames.

**Comment by mwcampbell** [comment]
@aevyrie Thanks for investigating this. What does LMB mean in this context?

I think the fact that it only happens with vsync enabled is significant. It tells me that when vsync is enabled, the application's main thread is blocked for a significant amount of time for each frame. This is bad for any accessibility client on macOS, and it appears that some OS component is acting as an accessibility client even when you're not running VoiceOver or another visible accessibility tool. The main thread needs to remain responsive to events as much as possible in order for accessibility clients to work quickly, because accessibility clients often have to make several IPC round trips to complete a single action. This isn't as severe an issue on Windows, because UI Automation provides a way to handle most accessibility requests away from the main thread, and AccessKit takes advantage of that. And it's not an issue at all in Unix desktop environments using AT-SPI.

**Comment by aevyrie** [comment]
@mwcampbell after doing some more debugging, I've edited my posts and created a new issue. While there may indeed be an access kit issue, there *also* seems to be a winit issue, specifically when left clicking and moving. By LMB, I mean left mouse button.

**Comment by mwcampbell** [comment]
I remain convinced that vsync is the culprit. The big question is whether it's practical, with wgpu, to do proper frame pacing without wgpu's `get_current_texture` blocking the main thread.

**Comment by geirsagberg** [comment]
Do you have any window managers installed, like Magnet? I noticed a strange lag on mouse events, especially dragging the window. Disabling Magnet made it go away.

**Comment by jmskov** [comment]
Can confirm that the suggestion by @geirsagberg to close Magnet removes the lag on M2 Macbook, 14.6.1.

**Comment by artursapek** [comment]
Just realized @geirsagberg posted an answer here. Nice. I'm going to close this.
